### ISSUE TITLE
Accessing records via CollectionProxy has poor performance compared to `target`

### ISSUE THREAD

**Issue by kevingriffin** [rails/rails#28656]
Our team discovered that iterating through a preloaded association has very poor performance compared to accessing the records directly via `target`.  Inspection into the issue shows lots of scope building that doesn't seem to be used as the cause.

### Steps to reproduce

Create a simple schema with associations. Load the database with a large number of records. Then, preload all of the associations and iterate over them in two ways: 1) with `name.each` and with `association(:name).target.each`. Measure the performance in both cases.

Here's a runnable example:

https://gist.github.com/kevingriffin/29d7e4774d210a41116c063dfd430de0

### Expected behavior
Traversing the records of a pre-loaded association (e.g. via `each`) should have similar performance to first calling `target` on the association to access an array of the records before iterating.

### Actual behavior
Using `each` on pre-loaded associations is many orders of magnitude slower than calling `target` to access, then iterate the array of records.

### System configuration
**Rails version**: 5.0.1

**Ruby version**: 2.3.3

**Comment by kamipo** [comment]
This is due to the instanciation cost of `CollectionProxy` is very expensive.
So the first time accessing is very slow. It was fixed in #25877.

current master:

```
CollectionProxies (1000000):   0.460000   0.010000   0.470000 (  0.461838)
Associations (1000000):   0.110000   0.000000   0.110000 (  0.116738)
```

**Comment by kevingriffin** [comment]
Thank you! Awaiting 5.1. :)
